### PR TITLE
Support eventual consistency in conformance tests

### PR DESCRIPTION
**Pull request by nathancoleman**
**What type of PR is this?**

/kind bug
(best effort ^, is there a better option?)

**What this PR does / why we need it**:
We discussed in the sig meeting recently that "accepted" (or "ready" in the future) can only really indicate that the controller has synced all necessary configuration and that routes will begin working "soon".

In order for the conformance tests to support this, they must accept that HTTP requests to new routes may fail for some period of time but should become eventually consistent.

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: conformance/utils/http/http.go [hunk 1]
```diff
@@ -72,8 +73,29 @@ func MakeRequestAndExpectResponse(t *testing.T, r roundtripper.RoundTripper, gwA
...
-	cReq, cRes, err := r.CaptureRoundTrip(req)
-	require.NoErrorf(t, err, "error making request")
+
+	var cReq *roundtripper.CapturedRequest
+	var cRes *roundtripper.CapturedResponse
+
+	// We expect eventual consistency once the resources are ready, so wait until
+	// we've made a successful request and gotten a response with desired status
+	// before making assertions on the response body.
+	require.Eventually(t, func() bool {
```
Comment (howardjohn) [inline conformance/utils/http/http.go:83]: Do we want to test eventually **consistent**? Technically a proxy doing a 50/50 200 or 404 passes currently. We could do a check that does backoff on failures, and on success retry immediately and ensure we get the "success" N times.

In Istio tests, we do all of ours this way fwiw and it seems to work pretty well. We do exponential backoff and ensure we get 3 successes consecutively

File: conformance/utils/http/http.go [hunk 2]
```diff
@@ -72,8 +73,29 @@ func MakeRequestAndExpectResponse(t *testing.T, r roundtripper.RoundTripper, gwA
...
+		var err error
+
+		cReq, cRes, err = r.CaptureRoundTrip(req)
+		if err != nil {
+			t.Log("request failed, not ready yet")
+			return false
+		}
+
+		if cRes.StatusCode != expected.StatusCode {
+			t.Log("response missing expected status, not ready yet")
```
Comment (howardjohn) [inline conformance/utils/http/http.go:93]: Can we log what we expect vs what we got?

File: conformance/utils/http/http.go [hunk 3]
```diff
@@ -72,8 +73,29 @@ func MakeRequestAndExpectResponse(t *testing.T, r roundtripper.RoundTripper, gwA
...
+			return false
+		}
+
+		return true
+	}, 60*time.Second, 1*time.Second, "error making request, never got expected status")
```
Comment (nathancoleman) [inline conformance/utils/http/http.go:98]: I think it makes sense to have this 60s cap be configurable, but I'm not sure at what granularity. Same cap for every test in the suite? Add as a func arg so every caller can specify their own?

Looking for some input from others here

File: conformance/utils/http/http.go (hunk 1 above)
Comment (nathancoleman) [reply conformance/utils/http/http.go:87]: @howardjohn if you would, please take a look at latest commits and let me know what ya think. Each test that needs to make assertions using this path now looks like this:

image

**Comment by evankanderson** [comment]
Any chance that we could instead require a conservative `Ready` status for the `HttpRoute`, rather than expecting each client to build in eventual consistency handling?

I ask because making the Gateway API eventually consistent (rather than conservatively `Ready`) means that every development team writing an end-to-end test through the Gateway API will need to add this logic to their tests and potentially have to debug the "was it a Gateway-induced eventual consistency failure or an application-level failure" when their tests fail or time out. As an end-user developer writing tests, being able to distinguish between "the Gateway controller didn't finish" and "my application fails for 30s at startup" is pretty useful.

**Comment by nathancoleman** [comment]
Hi @evankanderson ! There was some discussion about adding a `Ready` status to the `HTTPRoute`, and it sounds like that will be coming in the near term; however, I believe even then, things are expected to be eventually consistent due to the way these systems work.

The context from the discussion at yesterday's meeting might be helpful. It's up on YouTube [here](https://youtu.be/IwWuwGneXno?t=941) (linked to appropriate timestamp).

**Review by skriss — APPROVED** [review]
FYI I've been able to run the conformance tests using this PR against Contour's implementation successfully, whereas the tests in `master` where flaky for us. Thanks!

File: conformance/utils/http/http.go [hunk 4]
```diff
@@ -72,10 +80,37 @@ func MakeRequestAndExpectResponse(t *testing.T, r roundtripper.RoundTripper, gwA
...
+
+		if cRes.StatusCode != expected.StatusCode {
+			t.Logf("Expected response to have status %d but got %d, not ready yet", expected.StatusCode, cRes.StatusCode)
+			return false
+		}
+
+		return true
+	}, maxConsistencyPeriodPerRequest, 1*time.Second, "error making request, never got expected status")
+
+	// Once we've made a successful request and gotten a response with the
```
Comment (howardjohn) [inline conformance/utils/http/http.go:104]: Sorry to nitpick this a lot... a bit worried about the logic here. Typically what I have seen is that config changes will go from "all old config" to "mix of old and new" before arriving at "new config". So if we assert that once we see the 'new' state its immediately always 'new', we may see issues.

For example, if we have 2 pods implementing the gateway and 1 gets the new config first.

Here is how the Istio logic for this works: https://github.com/istio/istio/blob/30a866fdb5ad65eccde3dafe21d19f455addbd72/pkg/test/util/retry/retry.go#L182.

Basically we test that we get a success N times in a row -- but if we fail part way through the check we continue on retrying.

Also a bit of nitpick, but I am not sure we need the `consistency_check_%d` to show up in the test result? Doing the above would make that pretty hard to retain anyways

Comment (robscott) [reply conformance/utils/http/http.go:104]: I agree with the logic suggested here by @howardjohn. Maybe a fairly low timeout like 30 seconds to get 3 consecutive successful requests would work?

**Review by robscott — COMMENTED** [review]
Thanks for the work on this @nathancoleman! Agree with John's suggestion but otherwise this LGTM.

**Comment by robscott** [comment]
>Any chance that we could instead require a conservative Ready status for the HttpRoute, rather than expecting each client to build in eventual consistency handling?

@nathancoleman I really wish that was possible. That was my original intent here, but unfortunately I think it was a bit too optimistic. Many (maybe most) of the implementations of this API simply don't have a way to know when the config they've written has actually been picked up and implemented. If there are good ways around this, I'm very open to alternatives, because I agree that a reliable "ready" condition on Routes would be more helpful than a condition that indicated that config had been updated and should be ready to use soon.

**Comment by evankanderson** [comment]
Two thoughts:

1. Does it make sense to call this condition `Accepted` or `Programmed` rather than `Ready` to make the distinction clearer to users?
2. Conservative here could include adding a 30s delay after `Accepted`, assuming that's something like 2.5 stddev beyond the mean.

File: conformance/utils/http/http.go (hunk 4 above)
Comment (nathancoleman) [reply conformance/utils/http/http.go:104]: 👍 Not nitpicking IMO @howardjohn ! This makes a lot of sense and will help us avoid similar issues when we make the number of pods implementing the gateway configurable beyond `1` in the near future.

Comment (nathancoleman) [reply conformance/utils/http/http.go]: @howardjohn I've pushed some updates to handle intermittent success in the style of what you linked. Let me know what ya think!

You can see an example run w/ log output under **Run tests** [here](https://github.com/hashicorp/consul-api-gateway/runs/5793540213?check_suite_focus=true).

**Comment by nathancoleman** [comment]
@evankanderson @robscott I believe that future status additions for `HTTPRoute` -- "ready" vs. "accepted" vs. "programmed", etc., none of which exists today -- would be a good topic for the next kube-sig meeting on Monday, April 4.

Those evolutions in the Gateway API are only tangentially related to this PR. My goal with this change is just to unblock several implementations that are unable to run the conformance tests successfully today.

Would it make sense to you guys to move this discussion over to that venue? I think that would allow for more participation as well.

File: conformance/utils/http/http.go [hunk 5]
```diff
@@ -44,9 +45,15 @@ type ExpectedRequest struct {
 	Headers map[string]string
 }
 
-// MakeRequestAndExpectResponse makes a request with the given parameters and
-// verifies the response matches the provided ExpectedResponse.
-func MakeRequestAndExpectResponse(t *testing.T, r roundtripper.RoundTripper, gwAddr string, expected ExpectedResponse) {
+const maxConsistencyPeriodPerRequest = 60 * time.Second
```
Comment (robscott) [inline conformance/utils/http/http.go:48]: As I understand it, these tests require 3 consecutive successful requests to have occurred within this time period, and this time starts once a Route transitions to a status of accepted. If that's all correct, I think I'd prefer starting with a lower value of 30 seconds + renaming this to something like `maxTimeToConsistency`. Not sure on naming, but the `PerRequest` part throws me off in the current name.

File: conformance/utils/http/http.go [hunk 6]
```diff
@@ -44,9 +45,15 @@ type ExpectedRequest struct {
...
+const numConsistencyChecksPerRequest = 3
```
Comment (robscott) [inline conformance/utils/http/http.go:49]: As I understand this, this variable controls the number of consecutive successful requests we require for a test to be considered successful. If that's a correct interpretation, the current name doesn't clearly communicate that to me. Maybe something like `requiredConsecutiveSuccesses` or similar would be clearer. It's hard to come up with a perfectly descriptive name for either of these vars, so it may also make sense to add comments describing the purpose of each of them.

File: conformance/utils/http/http.go [hunk 7]
```diff
@@ -72,12 +79,49 @@ func MakeRequestAndExpectResponse(t *testing.T, r roundtripper.RoundTripper, gwA
...
+		cRes         *roundtripper.CapturedResponse
+		err          error
+		numSuccesses int
+	)
+
+	require.Eventually(t, func() bool {
+		cReq, cRes, err = r.CaptureRoundTrip(req)
+		if err != nil {
+			numSuccesses = 0
+			t.Logf("Request failed, not ready yet: %v", err.Error())
```
Comment (robscott) [inline conformance/utils/http/http.go:102]: Looking at the ways CaptureRoundTrip can return an error, I'm wondering if we should exit early with a test failure here like we were before (`require.NoErrorf(t, err, "error making request")`). For example, if the error we're getting is from `http.NewRequestWithContext` (https://github.com/kubernetes-sigs/gateway-api/blob/master/conformance/utils/roundtripper/roundtripper.go#L87), we're not going to have any more success retrying with the same params.

**Review by robscott — COMMENTED** [review]
Thanks @nathancoleman! A few more tiny nits but otherwise this LGTM.

File: conformance/utils/http/http.go (hunk 7 above)
Comment (nathancoleman) [reply conformance/utils/http/http.go:102]: `CaptureRoundTrip` will return an error in the event that Envoy has not stood up listeners on the gateway yet even if it's been "programmed", so I don't think we should fail the test at this point. The general progression we see in our conformance test runs is:

```log
# Envoy isn't ready yet, so this returns an error from CaptureRoundTrip
http.go:102: Request failed, not ready yet: Get "http://172.18.255.201/": dial tcp 172.18.255.201:80: connect: connection refused

http.go:114: Request has passed 1 times in a row of the desired 3, not ready yet
http.go:114: Request has passed 2 times in a row of the desired 3, not ready yet
http.go:118: Request has passed 3 times in a row of the desired 3, ready!
```
(taken from the **Run tests** step in our Actions workflow [here](https://github.com/hashicorp/consul-api-gateway/runs/5821189907?check_suite_focus=true))

File: conformance/utils/http/http.go (hunk 5 above)
Comment (nathancoleman) [reply conformance/utils/http/http.go:48]: The route transition happens outside the scope of the functions I've modified/added. The timer starts when the function is called essentially. The caller has already verified the status transition before calling `MakeRequestAndExpectEventuallyConsistentResponse`

File: conformance/utils/http/http.go (hunk 7 above)
Comment (robscott) [reply conformance/utils/http/http.go:109]: Good catch, that makes sense to me.

File: conformance/utils/http/http.go (hunk 6 above)
Comment (nathancoleman) [reply conformance/utils/http/http.go]: Renamed + added comments

**Comment by robscott** [comment]
Thanks @nathancoleman!

/lgtm
/approve
